### PR TITLE
App cert

### PR DESCRIPTION
**Pull request by obino**
This PRs allows to have one certificate per application. The internal (self-signed) certificate created at deploy time, it is still used for internal communication, but now each application can use its own certificate. The dashboard is still using the internal certificate.

### CODE REVIEW

**Comment by obino** [comment]
OCD build http://ocd.appscale.net:8080/job/Daily%20Build/6119/

File: AppController/lib/nginx.rb [hunk 1]
```diff
@@ -433,6 +442,41 @@ def self.add_service_location(service_name, service_host, service_port,
     Nginx.reload
   end
 
+  def self.ensure_certs_are_in_place(project_id=nil)
+    # If the project is nil, we'll set up the self-signed certs for the
+    # internal communication.
+    target_certs = ["#{NGINX_PATH}/mycert.pem", "#{NGINX_PATH}/mykey.pem"]
+    src_certs = ["#{Djinn::APPSCALE_CONFIG_DIR}/certs/mycert.pem",
+                 "#{Djinn::APPSCALE_CONFIG_DIR}/certs/mykey.pem"]
+    ret = false
```
Comment (cdonati) [inline AppController/lib/nginx.rb]: I think something like `certs_modified` would be more meaningful here to indicate what this function is returning.

File: AppController/lib/nginx.rb [hunk 2]
```diff
@@ -297,9 +301,14 @@ def self.write_fullproxy_version_config(version_key, http_port, https_port,
...
                             "appscale-#{version_key}.#{CONFIG_EXTENSION}")
 
-    # Let's reload and overwrite only if something changed.
+    # Let's reload and overwrite only if something changed, or new
+    # certificates have been installed.
     current = ''
     current = File.read(config_path) if File.exists?(config_path)
+
+    # Make sure we have the proper certificates in place.
+    current = '' if ensure_certs_are_in_place(project_id)
```
Comment (cdonati) [inline AppController/lib/nginx.rb]: Modifying `current` here just to make it different than `config` seems like a bit of a hack. I would recommend something like
```
certs_changed = ensure_certs_are_in_place(project_id)
if current != config || certs_changed
```

File: AppController/lib/nginx.rb [hunk 3]
```diff
@@ -433,6 +442,41 @@ def self.add_service_location(service_name, service_host, service_port,
...
+    # Validate and use the project specified certs for the project.
+    if !project_id.nil?
+      target_certs = ["#{NGINX_PATH}/#{project_id}.pem",
+                      "#{NGINX_PATH}/#{project_id}.key"]
+      new_src_certs = ["#{Djinn::APPSCALE_CONFIG_DIR}/certs/#{project_id}.pem",
+                       "#{Djinn::APPSCALE_CONFIG_DIR}/certs/#{project_id}.key"]
+      if File.exist?(new_src_certs[0]) && File.exist?(new_src_certs[1])
+        if system("openssl x509 -in #{new_src_certs[0]} -noout") &&
+            system("openssl rsa -in #{new_src_certs[1]} -noout")
+          src_certs = new_src_cert
```
Comment (cdonati) [inline AppController/lib/nginx.rb]: I'm guessing this should be `new_src_certs`.

**Comment by obino** [comment]
OCD build http://ocd.appscale.net:8080/job/Daily%20Build/6317/

**Pull request merged**